### DISCUSSION TITLE
Add global volume envelope ADSR parameters

### DISCUSSION THREAD

**Discussion by sound8** [FluidSynth/fluidsynth#1255] [Ideas]
Have seen some stuff mentioned about this here and there in the past but was wondering if it would be possible to add global volume envelope ADSR parameters.

Fluidsynth seems to offer tons of parameters to the developers who build from it but as far as I know it does not contain global parameters for this currently.

I have seen many Soundfont players in Windows (that are not Fluidsynth based) offer these global ADSR parameters, so it is possible to achieve with the Soundfont format.

In Linux especially I have noticed that Fluidsynth is the base for literally all software that provides Soundfont support.
Every GUI created with the software produces the same / similar interface that contains two effects chorus and reverb with four parameter knobs a piece.
Without this feature this means that the Linux platform will never have a Soundfont player that has these controls available.

If feel that it is needed can also add an on/off parameter to be able to turn this envelope on and off so that user does not have to set the four ADSR knobs back to achieve the same results.

The next challenge will be getting a developer of a Soundfont player built off of Fluidsynth to actually add the parameters to their GUI but at least they will be there if they agree to do so.
 
I apologize for bringing this topic up again.
Even if an answer is provided that the developers of Fluidsynth are not interested in this feature would be helpful to know.
 
Thank You to anyone working on the project.

**Comment by derselbst** [comment]
I've converted this to a discussion because your request is currently way too vague.

Firstly: Pls. describe your use case! So far, you have only argued: "Others have it so fluidsynth needs to have it as well." - I do not support this way of arguing.

Secondly: I have no idea, what these "global ADSR volume" parameters should behave like nor influence on a technical level nor add up with the already existing ADSR logic of the SF2 spec. Perhaps I could derive it from your use case if you had provided one.

And after that, we can think about if / how to implement or support this case.

**Comment by sound8** [comment]
@derselbst
Ok…. not the response I was expecting.
You seem to be rather upset.
I just thought your program was pretty great and I was just trying to make a suggestion on a Github page.
I was and am in no way trying to insult you or Fluidsynth program.
I apologize, Fluidsynth is a pretty in depth program so I thought what I was recommending was fairly self explanatory.
I just made an assumption because from my point of view anyone who can create a program like this I figured would know a great deal about soundfonts and all the different ways they are used.  I can explain further as requested.
You moved this post to the “discussions” section so….. lets discuss.
 
As far as I know, there are two general uses for soundfont players (but I am sure they can be used for all sorts of applications)
1.) They load up soundfonts and use / play them as is.
Programs send saved MIDI data found in files that are in .mid file format into the soundfont player which generates the sounds based on the soundfont loaded in the player.
So for example I have seen people online download soundtracks from say video games that contain a bunch of .mid tracks and they load them all up in MIDI Sequencer / DAW and each track gets assigned an instrument and all of them together generates the music from that video game.
 
2.) They are used as more of a sampler based instrument allowing users to play / create music with them.
User can load a SF2 into a player (Standalone or Plugin based such as VST or LV2) and play a MIDI keyboard and in realtime send MIDI data through the SF2 which generates sound realtime based on what is being played.
 
Although what I am requesting lends more to the 2nd item mentioned above, it could be used in both because generally the SF2 Player just needs MIDI data from somewhere.  (I personally care more about the 2nd item)
 
I have looked into Fluidsynth from time to time and am amazed by all the parameters / MIDI CC items made available.
My comment was regarding that with a program that offers so much depth I just can’t understand why developers of GUI (based on Fluidsynth) never allow access for these parameters in their GUIs
As mentioned they typically just show the two effects that come with Fluidsynth and not much else.
This could make many who have not looked into what the program can do, think that this is all there is, which is not the case.
So all the great work being done is not really being accessed or known that it could be accessed.
 
Okay… onto explaining what “Global” volume envelope ADSR does….
(I imagine you know what a Volume envelope ADSR does so I am not going to explain this unless you would like me to)
Currently the soundfont players created that are based on Fluidsynth do not let users alter the sound of the soundfont instruments.
Whatever the instruments are programmed to sound like that is it, the user does not have many options available in most GUI to alter anything else.
If they would want to change / alter the sounds they would have use a Soundfont editor to change the parameters of a soundfont this way.
One of those parameters would be volume envelope ASDR, but regardless any parameters that are changed would then have to be saved as a new instrument / copy.
These parameters cannot be adjusted in realtime by the user in the player itself.
What a “Global” volume envelope, that contains the parameters ADSR (attack, decay, sustain, release) would allow is for the user to be able to adjust these parameters to alter the sound in realtime, no editing, no need to save a new instrument, etc.  
This "Global" set of parameters applies to each note / audio sample within the soundfont instrument, any note that plays gets this envelope applied to it.
So for example if a soundfont has a rather abrubt end to its sounds when the MIDI note is over, the user could adjust the "release" parameter which would allow the note to play on a little after the release of the MIDI note is over.
Like how a string would if plucked.

If what I am saying does not make sense, you can think of it somewhat similar to the “legato / glide” parameters offered by Fluidsynth (which is another parameter that I wish developers of GUIs for Fluidsynth would make available in their GUIs)
When turned on this parameter allows the user the ability to alter the sound of the soundfont without having to actually edit it, it is done in realtime.
 
When I mentioned the fact that other soundfont players that are not Fluidsynth based you mentioned that you “do not support this way of arguing”.
No argument intended, the only reason I brought this up was to give you an example of what I was referring to and to also show that it could be a feature that is valid enough for other developers to make available, that this isn’t something I was just making up.
The only soundfont player that I have found that was based on Fluidsynth that has volume envelope ADSR built into its GUI is a player called “JuicySF” which is no longer being developed and the the Linux version was not maintained after a certain point.  With this program the developer ended up creating these parameters and added them to the GUI.
https://github.com/Birch-san/juicysfplugin
 
If there is anything that I can explain further, please let me know.
Hopefully something like this will be considered to be added to your software, if it is possible.  As mentioned if you can let me know either way.

**Comment by agraef** [comment]
Sorry, I accidentally dropped by and can't help wondering: What are you even arguing about? Those global envelope parameters are in the spec (see http://www.synthfont.com/SFSPEC21.PDF, page 48, parameters # 25 thru 30) and in the [fluid_gen_type enum](https://www.fluidsynth.org/api-1.x/gen_8h.html#ad17a24ae3b25f3b8cc5762f818eef9b4) , so they should be modifiable with [fluid_synth_set_gen()](https://www.fluidsynth.org/api-1.x/synth_8h.html#ae5d823189a618befba732ec679beac74), no? Or am I missing something here?

**Comment by derselbst** [comment]
> You seem to be rather upset.

Right. I am indeed not amused if someone bypasses the community by directly opening a feature-ticket, ignoring that it says "Feature requests should be discussed by the community, either on the mailing list or via a GitHub Discussion. Please bring up your ideas there before opening tickets." And even less amused when the first sentence I read is "Have seen some stuff mentioned about this here and there...".

---

Back to topic. All the parameters, like modulation, ADSR, vibrato, etc. can be manipulated in realtime by using modulators. This is the way the SoundFont specification wants people to do things. And this is what's written in our [wiki](https://github.com/FluidSynth/fluidsynth/wiki/FluidFeatures). E.g. you can set up a modulator for a particular instrument that allows you to e.g. let CC23 control the release time.

If you want this handling to be "global" you just need to add this modulator to all instruments in the SoundFont (with a SoundFont editor). This way, the realtime-manipulation would be even fully portable among SF2-compliant players. Alternatively, you can use [fluid_synth_add_default_mod()](https://www.fluidsynth.org/api/group__synthesis__params.html#gaa5b17b128a04237de73262a6e591be17) to do this programmatically when targeting fluidsynth specifically. I have done this to properly [play back N64 Tunes](https://github.com/derselbst/ANMP/blob/f335127b6f0a31f6da3f7074c98d0485edc7cfef/src/InputLibraryWrapper/FluidsynthWrapper.cpp#L219-L317), just [like Brich-san has done it for JuicySF](https://github.com/Birch-san/juicysfplugin/blob/c95eb5755aab634a22dfa778912facd136590d5c/Source/FluidSynthModel.cpp#L104-L177).

Or you can use the function mentioned by Albert. Though this would apply to MIDI channels and not instruments (tiny difference that will matter when rapidly changing instruments on channels).

Or you can use the synth's API to retrieve and manipulate individual voices.

Controls are possible on so many levels that I currently don't see what you are missing in fluidsynth, sry.

**Comment by mateusz** [comment]
Hey, in case someone will find this useful - I went through a similar rabbit-hole, looking for a simple command-line solution to get my MIDI piano's ADSR working with fluidsynth. I just wanted to sustain the notes a bit longer, which worked out of the box on all UI-based synths I tried.

I ended up getting the relevant piece out of JuicySFPlugin UI https://github.com/Birch-san/juicysfplugin (thank you!), and creating a small CLI wrapper like so https://github.com/mateusz/fluidadsr .
